src/abstract/operators.jl

```diff
@@ -17,98 +17,92 @@ function isless(m1::AbstractMonomial{V}, m2::AbstractMonomial{V}) where {V}
     end
 end
 
-function isless(t1::AbstractTerm, t2::AbstractTerm)
-    if monomial(t1) < monomial(t2)
-        true
-    elseif monomial(t1) == monomial(t2)
-        coefficient(t1) < coefficient(t2)
-    else
-        false
-    end
-end
-
-for op in [:+, :-, :(==)]
-    @eval $op(p1::AbstractPolynomialLike, p2::AbstractPolynomialLike) = $op(promote(p1, p2)...)
-    @eval $op(p::AbstractPolynomialLike, x) = $op(promote(p, x)...)
-    @eval $op(x, p::AbstractPolynomialLike) = $op(promote(x, p)...)
-end
+#function isless(t1::TypedTerm, t2::TypedTerm)
+#    if monomial(t1) < monomial(t2)
+#        true
+#    elseif monomial(t1) == monomial(t2)
+#        coefficient(t1) < coefficient(t2)
+#    else
+#        false
+#    end
+#end
 
 for op in [:+, :-]
-    @eval $op(p1::AbstractTermLike, p2::AbstractTermLike) = $op(convert(AbstractPolynomial, p1), convert(AbstractPolynomial, p2))
-    @eval $op(p1::AbstractPolynomial, p2::AbstractTermLike) = $op(p1, convert(AbstractPolynomial, p2))
-    @eval $op(p1::AbstractTermLike, p2::AbstractPolynomial) = $op(convert(AbstractPolynomial, p1), p2)
+    @eval $op(p1::TypedTermLike, p2::TypedTermLike) = $op(polynomial(p1), polynomial(p2))
+    @eval $op(p1::TypedPolynomial, p2::TypedTermLike) = $op(p1, polynomial(p2))
+    @eval $op(p1::TypedTermLike, p2::TypedPolynomial) = $op(polynomial(p1), p2)
 end
 
-(+)(p1::AbstractPolynomial, p2::AbstractPolynomial) = convert(AbstractPolynomial, jointerms(terms(p1), terms(p2)))
+(+)(p1::TypedPolynomial, p2::TypedPolynomial) = polynomial(jointerms(terms(p1), terms(p2)))
 
-(-)(t::AbstractTermLike) = -1 * t
-(-)(p1::AbstractPolynomial, p2::AbstractPolynomial) = convert(AbstractPolynomial, jointerms(terms(p1), (-).(terms(p2))))
+#(-)(t::TypedTermLike) = -1 * t
+(-)(p1::TypedPolynomial, p2::TypedPolynomial) = polynomial(jointerms(terms(p1), (-).(terms(p2))))
 
 
-@pure (==)(::AbstractVariable{N}, ::AbstractVariable{N}) where {N} = true
-@pure (==)(::AbstractVariable, ::AbstractVariable) = false
-(==)(m1::AbstractMonomial{V}, m2::AbstractMonomial{V}) where {V} = exponents(m1) == exponents(m2)
-function (==)(t1::AbstractTerm, t2::AbstractTerm)
-    c1 = coefficient(t1)
-    c2 = coefficient(t2)
-    if iszero(c1) && iszero(c2)
-        true
-    else
-        c1 == c2 && monomial(t1) == monomial(t2)
-    end
-end
-(==)(::AbstractPolynomialLike, ::Void) = false
-(==)(::Void, ::AbstractPolynomialLike) = false
+@pure (==)(::TypedVariable{N}, ::TypedVariable{N}) where {N} = true
```

Comment:
No, this is specific to TypedPolynomials and StaticPolynomials. We could have an `AbstractNamedVariable{N} <: AbstractVariable` in MultivariatePolynomials if we want to move this there.